# <center> **Kaggle’s Spaceship Titanic Competition**
# <center> **Imputing Missing Values**

# **Libraries**

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as mpl
import matplotlib.pyplot as plt

import functions
import importlib
importlib.reload(functions)

from sklearn.preprocessing import LabelEncoder

import warnings

# **Load Data**

In [3]:
data = pd.read_csv(
    r"C:\Users\Dell\Documents\AI\Titanic\Data\data.csv",
    index_col=False
)

random_state = 101
target = 'Transported'

# data = data.drop([target], axis=1)

# **Display Parameters**

In [22]:
size = 20

params = {
    "font.family": "Times New Roman",
    "font.size": size,
    "axes.labelsize": size,
    "xtick.labelsize": size * 0.75,
    "ytick.labelsize": size * 0.75,
    "figure.titlesize": size * 1.5,
    "axes.titlesize": size * 1.5,
    "axes.titlepad": size,
    "axes.labelpad": size - 10,
    "lines.linewidth": 2,
    "axes.spines.top": False,
    "axes.spines.right": False,
    "axes.spines.left": False,
    "axes.spines.bottom": False,
    "legend.fontsize": size,
    "figure.figsize": (10, 6),
}

# **Imputing Missing Values**

In [4]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
HomePlanet             288               2.22
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
Destination            274               2.11
VIP                    296               2.28
Transported           4277              32.98

## **Home Planet**

### **Impute Home Planet from Group**

In [5]:
unique_homeplanets_per_group = data.groupby('Group')['HomePlanet'].nunique()
groups_with_multiple_planets = unique_homeplanets_per_group[unique_homeplanets_per_group > 1]

print(groups_with_multiple_planets)

Series([], Name: HomePlanet, dtype: int64)


In [6]:
group_homeplanet_map = data.groupby('Group')['HomePlanet'].apply(lambda x: x.mode()[0] if not x.mode().empty else None)

def fill_missing_homeplanet(row):
    if pd.isna(row['HomePlanet']):
        return group_homeplanet_map.get(row['Group'], row['HomePlanet'])
    else:
        return row['HomePlanet']

data['HomePlanet'] = data.apply(fill_missing_homeplanet, axis=1)

In [7]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
HomePlanet             157               1.21
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
Destination            274               2.11
VIP                    296               2.28
Transported           4277              32.98

### **Impute Home Planet from Last Name**

In [8]:
lastname_homeplanet_map = data.groupby('LastName')['HomePlanet'].apply(lambda x: x.mode()[0] if not x.mode().empty else None)

def fill_missing_homeplanet_by_lastname(row):
    if pd.isna(row['HomePlanet']):
        return lastname_homeplanet_map.get(row['LastName'], row['HomePlanet'])
    else:
        return row['HomePlanet']

data['HomePlanet'] = data.apply(fill_missing_homeplanet_by_lastname, axis=1)

In [9]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
HomePlanet               7               0.05
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
Destination            274               2.11
VIP                    296               2.28
Transported           4277              32.98

### **Impute Home Planet from Mode**

In [10]:
homeplanet_mode = data['HomePlanet'].mode()[0]
print(homeplanet_mode)

Earth


In [11]:
mode_value = data['HomePlanet'].mode()

if not mode_value.empty:
    data['HomePlanet'] = data['HomePlanet'].fillna(mode_value[0])

In [12]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
Destination            274               2.11
VIP                    296               2.28
Transported           4277              32.98

### **Insights**

>  **0. The Code:** Grouping HomePlanet by Group/LastName. Retrieving the most common HomePlanet (mode) based on Group/LastName</BR>
>  **1. Grouping by Group:** Starting with 288 missing values. Grouping by 'Group' to find the most common (or only) HomePlanet. Fill missing HomePlanet values based on Group's identified home planet. End with 157 missing values. </BR>
>  **2. Grouping by Last Name:** Starting with 157 missing values. Grouping by 'LastName' to find the most common (or only) HomePlanet. Fill missing HomePlanet values based on last name's identified home planet. End with 7 missing values. </BR>
>  **3. Mode:** Start with 7 missing values. Fill missing values by mode of HomePlanet, which is Earth. End with 0 missing values.

## **Destination**

### **Impute Destination from Group**

In [18]:
unique_destinationplanets_per_group = data.groupby('Group')['Destination'].nunique()
groups_with_multiple_destinations = unique_destinationplanets_per_group[unique_destinationplanets_per_group > 1]

groups_with_multiple_destinations.sort_values(ascending=True)

Group
8       2
5856    2
5863    2
5876    2
5885    2
       ..
6959    3
1814    3
5911    3
6787    3
1304    3
Name: Destination, Length: 1070, dtype: int64

In [397]:
group_destinationplanet_map = data.groupby('Group')['Destination'].apply(lambda x: x.mode()[0] if not x.mode().empty else None)

def fill_missing_destinationplanet(row):
    if pd.isna(row['Destination']):
        return group_destinationplanet_map.get(row['Group'], row['Destination'])
    else:
        return row['Destination']

data['Destination'] = data.apply(fill_missing_destinationplanet, axis=1)

In [398]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
Destination            154               1.19
VIP                    296               2.28
Transported           4277              32.98

### **Impute Destination from Last Name**

In [399]:
lastname_destinationplanet_map = data.groupby('LastName')['Destination'].apply(lambda x: x.mode()[0] if not x.mode().empty else None)

def fill_missing_destinationplanet_by_lastname(row):
    if pd.isna(row['Destination']):
        return lastname_destinationplanet_map.get(row['LastName'], row['Destination'])
    else:
        return row['Destination']

data['Destination'] = data.apply(fill_missing_destinationplanet_by_lastname, axis=1)

In [400]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
Destination              4               0.03
VIP                    296               2.28
Transported           4277              32.98

### **Impute Destination from Mode**

In [401]:
desinationplanet_mode = data['Destination'].mode()[0]
print(desinationplanet_mode)

TRAPPIST-1e


In [402]:
data['Destination'] = data['Destination'].fillna(data['Destination'].mode()[0])

In [403]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

>  **0. The Code:** Grouping Destination by Group/LastName. Retrieving the most common Destination (mode) based on Group/LastName</BR>
>  **1. Grouping by Group:** Starting with 274 missing values. Grouping by 'Group' to find the most common (or only) destination. Fill missing values based on Group's identified destination. End with 154 missing values. </BR>
>  **2. Grouping by Last Name:** Starting with 154 missing values. Grouping by 'LastName' to find the most common (or only) destination. Fill missing values based on last name's identified destination. End with 4 missing values. </BR>
>  **3. Mode:** Start with 4 missing values. Fill missing values by mode, which is TRAPPIST-1e. End with 0 missing values.

## **Cabin Deck**

### **Cabin Deck and Home Planet**

In [404]:
data.pivot_table(index='CabinDeck', columns='HomePlanet', aggfunc='size', fill_value=0)

HomePlanet  Earth  Europa  Mars
CabinDeck                      
A               1     353     0
B               0    1141     0
C               1    1101     0
D               3     306   411
E             597     203   523
F            2480       0  1759
G            3781       0     0
T               0      11     0

### **Impute CabinDeck G from Earth**

In [405]:
data.loc[(data['HomePlanet'] == 'Earth') & (data['CabinDeck'].isnull()), 'CabinDeck'] = 'G'

In [406]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinDeck              138               1.06
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

### **Impute CabinDeck from Last Name**

In [407]:
lastname_cabindeck_map = data.groupby('LastName')['CabinDeck'].apply(lambda x: x.mode()[0] if not x.mode().empty else None)

def fill_missing_cabindeck_by_lastname(row):
    if pd.isna(row['CabinDeck']):
        return lastname_homeplanet_map.get(row['LastName'], row['CabinDeck'])
    else:
        return row['CabinDeck']

data['CabinDeck'] = data.apply(fill_missing_cabindeck_by_lastname, axis=1)

In [408]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinNumber            299               2.31
CabinSide              299               2.31
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

>  **1. Decks and Home Planets:** Starting with 299 missing values. Passengers on decks A (Except 1 passenger), B, C (Except 1 passenger), T came from Europa. Passengers on deck G came from Earth. Passengers on decks D, E or F came from multiple planets. </BR>
>  **2. Earth and Cabin Deck G:** Fill missing CabinDeck 'G' where HomePlanet is 'Earth'. End with 138 missing values. </BR> 
>  **3. Grouping by Last Name:** Starting with 138 missing values. Grouping by 'LastName' to find the most common (or only) cabin deck. Fill missing values based on last name's identified cabin deck. End with 0 missing values. </BR>

## **Cabin Side**

In [409]:
unique_cabinside_per_group = data[data['GroupSize']>1].groupby('Group')['CabinSide'].nunique()
grouped_by_unique_count = unique_cabinside_per_group.value_counts().sort_index()

grouped_by_unique_count = grouped_by_unique_count.to_frame().reset_index()

### **Cabin Side from Group**

In [410]:
group_cabinside = data[data['GroupSize'] > 1].groupby(['Group', 'CabinSide']).size().unstack(fill_value=0)

missing_cabinside_mask = data['CabinSide'].isna() & data['Group'].isin(group_cabinside.index)

data.loc[missing_cabinside_mask, 'CabinSide'] = data.loc[missing_cabinside_mask, 'Group'].map(group_cabinside.idxmax(axis=1))

In [411]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinNumber            299               2.31
CabinSide              162               1.25
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

### **Cabin Side from Last Name**

In [412]:
group_cabinside = data[data['GroupSize'] > 1].groupby(['LastName', 'CabinSide']).size().unstack(fill_value=0)

missing_cabinside_mask = data['CabinSide'].isna() & data['LastName'].isin(group_cabinside.index)

data.loc[missing_cabinside_mask, 'CabinSide'] = data.loc[missing_cabinside_mask, 'LastName'].map(group_cabinside.idxmax(axis=1))

In [413]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinNumber            299               2.31
CabinSide               65               0.50
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

### **Replace Cabin Side Missing Values by Letter U**

In [414]:
# Missing values before
CS_bef=data['CabinSide'].isna().sum()

# Fill remaining missing values with outlier
data.loc[data['CabinSide'].isna(),'CabinSide']='U'

In [415]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinNumber            299               2.31
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

### **Insights**

>  **1. Grouping by Group:** Starting with 299 missing values. Everyone in the same group is also on the same cabin side. Grouping by 'Group' to find the most common (or only) cabin side. Fill missing values based on Group's identified cabin side. End with 162 missing values. </BR>
>  **2. Grouping by Last Name:** Starting with 162 missing values. Grouping by 'LastName' to find the most common (or only) cabin side. Fill missing values based on last name's identified destination. End with 65 missing values. </BR>
>  **3. Fill by U (Unknown):** Starting with 65 missing values. They will be replaced with letter U (Unknown). This is because we really don't know which one of the two (balanced) sides we should assign. End with 0 missing values.

## **Cabin Number**

In [418]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinNumber            299               2.31
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

In [419]:
data['CabinNumber'] = data.groupby(['Group'])['CabinNumber'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else 'Unknown'))

In [420]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CryoSleep              310               2.39
VIP                    296               2.28
Transported           4277              32.98

### **Insights**

>  **1. Starting Missing Values:** Starting with 299 missing values.</BR>
>  **2. Linear Regrerssion:** Extrapolate the missing cabin numbers using mode. End with 0 missing values.

## **CryoSleep**

In [26]:
pd.set_option('future.no_silent_downcasting', True)

In [27]:
data['CryoSleep'] = data['CryoSleep'].fillna(False).astype(bool)

In [28]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
Destination            274               2.11
VIP                    296               2.28
Transported           4277              32.98

### **Insights**

>  **1. Starting Missing Values:** Starting with 310 missing values.</BR>
>  **2. CryoSleep == False:** Imputed Missing values with False since most passengers were False. </BR>
>  **3. End Missing Values:** End with 0 missing values.

## **VIP**

In [29]:
data['VIP'] = data['VIP'].fillna(False).astype(bool)

In [30]:
functions.MissingValues(data)

NumberMissing  PercentageMissing
CabinDeck              299               2.31
CabinNumber            299               2.31
CabinSide              299               2.31
Destination            274               2.11
Transported           4277              32.98

### **Insights**

>  **1. Starting Missing Values:** Starting with 3296 missing values.</BR>
>  **2. VIP == False:** Imputed Missing values with False since most passengers were False. </BR>
>  **3. End Missing Values:** End with 0 missing values.

## **Save Dataset**

In [426]:
data.to_csv(r"C:\Users\Dell\Documents\AI\Titanic\Data\Data\data.csv", index=False)